fe/src/main/java/org/apache/doris/qe/Coordinator.java

```diff
@@ -848,18 +848,39 @@ private void computeFragmentHosts() throws Exception {
             // be BE that fragment's scannode locates,  avoid less data.
             // chenhao added
             boolean hasUnionNode = containsUnionNode(fragment.getPlanRoot());
+
             if (!(leftMostNode instanceof ScanNode) && !hasUnionNode) {
                 // there is no leftmost scan; we assign the same hosts as those of our
                 // leftmost input fragment (so that a partitioned aggregation
                 // fragment runs on the hosts that provide the input data)
                 PlanFragmentId inputFragmentIdx =
-                    fragments.get(i).getChild(0).getFragmentId();
+                        fragments.get(i).getChild(0).getFragmentId();
                 // AddAll() soft copy()
-                for (FInstanceExecParam execParams 
-                        : fragmentExecParamsMap.get(inputFragmentIdx).instanceExecParams) {
-                    FInstanceExecParam instanceParam = new FInstanceExecParam(null, execParams.host, 
-                            0, params);
-                    params.instanceExecParams.add(instanceParam);
+                int doris_exchange_instances= -1;
+                if (ConnectContext.get() != null && ConnectContext.get().getSessionVariable() != null) {
+                    doris_exchange_instances = ConnectContext.get().getSessionVariable().getDorisExchangeInstances();
+                }
+                if (doris_exchange_instances > 0 && fragmentExecParamsMap.get(inputFragmentIdx).instanceExecParams.size() > doris_exchange_instances) {
+                    // random select some instance
+                    List<TNetworkAddress> hosts = Lists.newArrayList();
+                    Set<String> cache = new HashSet<String>();
+                    for (FInstanceExecParam execParams: fragmentExecParamsMap.get(inputFragmentIdx).instanceExecParams) {
+                        String hostPort = execParams.host.getHostname() + execParams.host.getPort();
+                        if (!cache.contains(hostPort)) {
```

Comment:
can you write some comment for this judge? In which case there can be same hostPort?